document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -388,6 +399,52 @@ public CloseableIterator<Document> aggregate(
     return executeQueryV1(query);
   }
 
+  @Override
+  public Optional<Document> update(
+      final org.hypertrace.core.documentstore.query.Query query,
+      final java.util.Collection<SubDocumentUpdate> updates)
+      throws IOException {
+
+    try (final Connection connection = client.getNewConnection()) {
```

Comment:
We are using a fixed set of static connections right now. And so, we are reusing the same connection till it is valid. I think,  if we can reuse the same connecction like any other API - getConnection() and reset connection back to `connection.setAutoCommit(true)` at the end. That should work fine, what do you think?
cc: @saxenakshitiz

Reply:
If we use the same connection and keep calling the `setAutoCommit()`, then the **updates can be lost** (since we don't control concurrency in this library).

1. Thread 1 calls the new `update` method and we disable auto-commit for the shared connection
2. Thread 1 completes the operation and commits
3. Thread 1 is interrupted (before enabling auto-commit). Thread 2 calls the existing `upsert` method and (since we haven't yet enabled auto-commit and the commit is already complete), this upsert would be lost
4. Thread 1 resumes and enables the auto-commit

Another issue I see is incorrect aborting. Consider the scenario.
1. Thread 1 calls the new `update` method and we disable auto-commit for the shared connection
2. Thread 1 is interrupted. Thread 2 calls the existing `upsert` method (and is not committed)
3. Thread 1 resumes, but fails. We abort the transaction (including the `upsert` done by Thread 2).


A proper solution would be to use a connection pool. One could consider using two connections: one with auto-commit on and the other with auto-commit off. But, that wouldn't scale either because.
1. Thread 1 calls the new `update` method
2. Thread 2 calls the new `update` method
3. Thread 1 commits (at a point where thread 2 has not completed executing).

The same issue will persist with abort as well.

Reply:
As discussed, lets see if we can use a separate connection pool for APIs which need explicit commit.